### ISSUE TITLE
nn.BatchNorm1d seems not calculates correctly at .eval() mode.

### ISSUE THREAD

**Issue by Zhangyanbo** [pytorch/pytorch#100048]
### 🐛 Describe the bug

`nn.BatchNorm1d` seems not calculates correctly at `.eval()` mode.

> **Note:** I know it may have different output of BN at train / eval mode. But I designed a case that they should be the same.

Reproduce:

```python
# make a chain of BN to show increase the differences
bnlist2 = nn.Sequential(*[nn.BatchNorm1d(2, eps=0, momentum=0.01, affine=False) for _ in range(20)])
bnlist2.train()

# setup a random x
x = torch.randn(64, 2) * 2 - 1

# running 2048 times to make sure the `running_means` and `running_vars` be stable.
for _ in range(2048):
    y = bnlist2(x)


# calculate the output at different mode
bnlist2.train()
y = bnlist2(x)

bnlist2.eval()
y_eval = bnlist2(x)
```

Output:
```text
y[:5] = 
tensor([[ 0.3988,  0.4764],
        [ 0.7844, -0.2870],
        [-2.2525,  1.4318],
        [ 1.0413, -1.1295],
        [ 1.3380,  0.8300]])

y_eval[:5] = 
tensor([[ 0.3407,  0.4070],
        [ 0.6701, -0.2452],
        [-1.9243,  1.2232],
        [ 0.8896, -0.9650],
        [ 1.1431,  0.7091]])
```

Then, I manually compute the BN result by:

```python
mean = x.mean(dim=0)
var = x.var(dim=0, unbiased=False)

((x - mean) / (var + 0).sqrt())[:5]
```

And I got:

```text
tensor([[ 0.3988,  0.4764],
        [ 0.7844, -0.2870],
        [-2.2525,  1.4318],
        [ 1.0413, -1.1295],
        [ 1.3380,  0.8300]])
```

This is the same as the train mode, but different from the eval mode. I tried to set the model to be `.double()`, but it does not helps.

### Versions

PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.26.2
Libc version: N/A

Python version: 3.8.12 | packaged by conda-forge | (default, Jan 30 2022, 23:13:24)  [Clang 11.1.0 ] (64-bit runtime)
Python platform: macOS-13.2.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] pytorch-lightning==1.6.3
[pip3] torch==2.0.0
[pip3] torch-cluster==1.6.0
[pip3] torch-geometric==2.0.4
[pip3] torch-scatter==2.0.9
[pip3] torch-sparse==0.6.13
[pip3] torch-spline-conv==1.2.1
[pip3] torchaudio==0.13.1
[pip3] torchmetrics==0.8.2
[pip3] torchvision==0.14.1
[conda] numpy                     1.21.6           py38hf29d37f_0    conda-forge
[conda] pytorch-lightning         1.6.3                    pypi_0    pypi
[conda] torch                     2.0.0                    pypi_0    pypi
[conda] torch-cluster             1.6.0                    pypi_0    pypi
[conda] torch-geometric           2.0.4                    pypi_0    pypi
[conda] torch-scatter             2.0.9                    pypi_0    pypi
[conda] torch-sparse              0.6.13                   pypi_0    pypi
[conda] torch-spline-conv         1.2.1                    pypi_0    pypi
[conda] torchaudio                0.13.1                 py38_cpu    pytorch
[conda] torchmetrics              0.8.2                    pypi_0    pypi
[conda] torchvision               0.14.1                 py38_cpu    pytorch

cc @ezyang @gchanan @zou3519 @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by albanD** [comment]
High pri to check if it is a fp precision problem (https://pytorch.org/docs/stable/notes/numerical_accuracy.html) or a bug.

**Comment by mikaylagawarecki** [comment]
Hey,

I modified your script a little (decreased it to just 1 BatchNorm in the dict)

```
import torch
from torch import nn

torch.manual_seed(5)

# make a chain of BN to show increase the differences
bnlist2 = nn.Sequential(*[nn.BatchNorm1d(2, eps=0, momentum=0.01, affine=False) for _ in range(1)])
bnlist2.train()

# setup a random x
x = torch.randn(64, 2) * 2 - 1

# running 2048 times to make sure the `running_means` and `running_vars` be stable.
for _ in range(2048):
    y = bnlist2(x)


# calculate the output at different mode
bnlist2.train()
y = bnlist2(x)

print(f"y[:5]\n{y[:5]}")

bnlist2.eval()
y_eval = bnlist2(x)
print(f"y_eval[:5]\n{y_eval[:5]}")

mean = x.mean(dim=0)
var_unbiased = x.var(dim=0, unbiased=True)
var_biased = x.var(dim=0, unbiased=False)
print("var (unbiased=True)")
print(((x - mean) / (var_unbiased + 0).sqrt())[:5])
print("var (unbiased=False)")
print(((x - mean) / (var_biased + 0).sqrt())[:5])
```

And got 

```
y[:5]
tensor([[ 1.5184,  0.6304],
        [-1.7317, -1.6364],
        [ 1.6803, -0.1736],
        [-0.2858, -1.3669],
        [ 0.1910,  0.0813]])
y_eval[:5]
tensor([[ 1.5065,  0.6255],
        [-1.7181, -1.6235],
        [ 1.6671, -0.1722],
        [-0.2836, -1.3561],
        [ 0.1895,  0.0807]])
var (unbiased=True)
tensor([[ 1.5065,  0.6255],
        [-1.7181, -1.6235],
        [ 1.6671, -0.1722],
        [-0.2836, -1.3561],
        [ 0.1895,  0.0807]])
var (unbiased=False)
tensor([[ 1.5184,  0.6304],
        [-1.7317, -1.6364],
        [ 1.6803, -0.1736],
        [-0.2858, -1.3669],
        [ 0.1910,  0.0813]])   
```

Seems like in eval mode, the unbiased variance is used, whereas in train mode the biased variance is used, which is causing the discrepancy

**Comment by Zhangyanbo** [comment]
@MikaylaGawarecki, thank you for your helpful response. Do you happen to know if this difference is intentional or not? This small variation can have a significant impact when I stack it with numerous layers, particularly when working with flow-based models that usually require a lot of layers.

**Comment by mikaylagawarecki** [comment]
[This comment](https://github.com/pytorch/pytorch/issues/77427#issuecomment-1126737756) has a more nuanced discussion of the above. Removing high priority since this behavior is expected. However, actionable to document this! 

The [previous PR](https://github.com/pytorch/pytorch/pull/97974) only updated this for BatchNorm2d but we should do the same for BatchNorm1/3d, @Zhangyanbo would you be interested in sending this fix? :)

**Comment by Zhangyanbo** [comment]
@mikaylagawarecki Thanks! Let me try to do this document fix :)
